File: localrag/providers/ollama.py
"""Local Ollama provider: chat via /api/chat, embeddings via /api/embeddings."""

from __future__ import annotations

from typing import List

import requests

from ..config import Config


class OllamaProvider:
    name = "ollama"

    def __init__(self, config: Config) -> None:
        self.url = config.ollama_url
        self.model = config.ollama_model
        self.embed_model = config.ollama_embed_model

    def is_available(self) -> bool:
        try:
            requests.get(f"{self.url}/api/tags", timeout=2).raise_for_status()
            return True
        except Exception:
            return False

    def chat(self, system: str, user: str) -> str:
        resp = requests.post(
            f"{self.url}/api/chat",
            json={
                "model": self.model,
                "stream": False,
                "messages": [
                    {"role": "system", "content": system},
                    {"role": "user", "content": user},
                ],
            },
            timeout=180,
        )
        resp.raise_for_status()
        return resp.json()["message"]["content"].strip()

    def embed(self, texts: List[str]) -> List[List[float]]:
        vectors: List[List[float]] = []
        for text in texts:
            resp = requests.post(
                f"{self.url}/api/embeddings",
                json={"model": self.embed_model, "prompt": text},
                timeout=120,
            )
            resp.raise_for_status()
            vectors.append(resp.json()["embedding"])
        return vectors

File: node/lesson-1/src/providers/ollama.js
// Local Ollama provider: chat via /api/chat. Mirrors localrag/providers/ollama.py.
// Uses the Node 18+ built-in fetch. (Embeddings are out of scope in this port.)

export class OllamaProvider {
  constructor(config) {
    this.name = "ollama";
    this.url = config.ollamaUrl;
    this.model = config.ollamaModel;
    this.embedModel = config.ollamaEmbedModel;
  }

  async isAvailable() {
    try {
      const resp = await fetch(`${this.url}/api/tags`, {
        signal: AbortSignal.timeout(2000),
      });
      return resp.ok;
    } catch {
      return false;
    }
  }

  async chat(system, user) {
    const resp = await fetch(`${this.url}/api/chat`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        model: this.model,
        stream: false,
        messages: [
          { role: "system", content: system },
          { role: "user", content: user },
        ],
      }),
      signal: AbortSignal.timeout(180000),
    });
    if (!resp.ok) {
      throw new Error(`ollama /api/chat returned ${resp.status}: ${await resp.text()}`);
    }
    const data = await resp.json();
    return data.message.content.trim();
  }
}
